File: kotlin/53.maximum-subarray.kt
/*
 * @lc app=leetcode id=53 lang=kotlin
 *
 * [53] Maximum Subarray
 */

// @lc code=start
class Solution {
    fun maxSubArray(nums: IntArray): Int {
        if (nums.isEmpty()) return 0
        var max = nums[0]
        var cur = nums[0]
        for (i in 1 until nums.size) {
            cur = maxOf(nums[i], cur + nums[i])
            max = maxOf(max, cur)
        }
        return max
    }
}
// @lc code=end

File: 53.maximum-subarray.go
/*
 * @lc app=leetcode id=53 lang=golang
 *
 * [53] Maximum Subarray
 */

// @lc code=start
func maxSubArray(nums []int) int {
    if(len(nums) == 0) {
		return 0
	}

	max := nums[0]
	cur := nums[0]

	for i := 1; i < len(nums); i++ {
		cur = maxInt(nums[i], cur + nums[i])
		max = maxInt(max, cur)
	}

	return max
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// @lc code=end

